File: pocketbase/pb_migrations/1738223400_initial_schema.js
/// <reference path="../pb_data/types.d.ts" />
migrate((db) => {
  const dao = new Dao(db);

  // 创建 tags 表
  const tagsCollection = new Collection({
    id: "tags_collection_id",
    name: "tags",
    type: "base",
    schema: [
      {
        name: "name",
        type: "text",
        required: true,
        options: {
          min: 1,
          max: 50,
        },
      },
      {
        name: "slug",
        type: "text",
        required: true,
        options: {
          min: 1,
          max: 50,
          pattern: "^[a-z0-9]+(?:-[a-z0-9]+)*$",
        },
      },
      {
        name: "color",
        type: "text",
        required: false,
        options: {
          min: 0,
          max: 7,
        },
      },
      {
        name: "icon",
        type: "text",
        required: false,
        options: {
          min: 0,
          max: 50,
        },
      },
      {
        name: "order",
        type: "number",
        required: false,
        options: {
          min: 0,
          noDecimal: true,
        },
      },
    ],
    indexes: [
      "CREATE UNIQUE INDEX idx_tags_name ON tags (name)",
      "CREATE UNIQUE INDEX idx_tags_slug ON tags (slug)",
    ],
    listRule: "",
    viewRule: "",
    createRule: null,
    updateRule: null,
    deleteRule: null,
  });

  dao.saveCollection(tagsCollection);

  // 创建 changelogs 表
  const changelogsCollection = new Collection({
    id: "changelogs_collection_id",
    name: "changelogs",
    type: "base",
    schema: [
      {
        name: "title",
        type: "text",
        required: true,
        options: {
          min: 1,
          max: 200,
        },
      },
      {
        name: "description",
        type: "editor",
        required: true,
        options: {
          convertUrls: false,
        },
      },
      {
        name: "version",
        type: "text",
        required: true,
        options: {
          min: 1,
          max: 50,
        },
      },
      {
        name: "date",
        type: "date",
        required: true,
        options: {
          min: "",
          max: "",
        },
      },
      {
        name: "tags",
        type: "relation",
        required: false,
        options: {
          collectionId: "tags_collection_id",
          cascadeDelete: false,
          minSelect: null,
          maxSelect: null,
          displayFields: ["name"],
        },
      },
    ],
    indexes: [
      "CREATE INDEX idx_changelogs_date ON changelogs (date DESC)",
      "CREATE INDEX idx_changelogs_version ON changelogs (version)",
    ],
    listRule: "",
    viewRule: "",
    createRule: null,
    updateRule: null,
    deleteRule: null,
  });

  dao.saveCollection(changelogsCollection);

  // 创建 site_config 表
  const siteConfigCollection = new Collection({
    id: "site_config_collection_id",
    name: "site_config",
    type: "base",
    schema: [
      {
        name: "site_title",
        type: "text",
        required: true,
        options: {
          min: 1,
          max: 100,
        },
      },
      {
        name: "site_description",
        type: "text",
        required: true,
        options: {
          min: 1,
          max: 500,
        },
      },
      {
        name: "logo_url",
        type: "file",
        required: false,
        options: {
          maxSelect: 1,
          maxSize: 5242880,
          mimeTypes: ["image/jpeg", "image/png", "image/svg+xml", "image/gif", "image/webp"],
          thumbs: ["100x100"],
        },
      },
      {
        name: "primary_color",
        type: "text",
        required: false,
        options: {
          min: 0,
          max: 7,
          pattern: "^#[0-9A-Fa-f]{6}$",
        },
      },
    ],
    indexes: [],
    listRule: "",
    viewRule: "",
    createRule: null,
    updateRule: null,
    deleteRule: null,
  });

  dao.saveCollection(siteConfigCollection);

  // 插入初始标签数据
  const initialTags = [
    {
      name: "New Feature",
      slug: "new-feature",
      color: "#3B82F6",
      icon: "Sparkles",
      order: 1,
    },
    {
      name: "Bug Fix",
      slug: "bug-fix",
      color: "#EF4444",
      icon: "Bug",
      order: 2,
    },
    {
      name: "Improvement",
      slug: "improvement",
      color: "#10B981",
      icon: "TrendingUp",
      order: 3,
    },
    {
      name: "Security Update",
      slug: "security",
      color: "#F59E0B",
      icon: "Shield",
      order: 4,
    },
  ];

  initialTags.forEach((tag) => {
    const record = new Record(tagsCollection);
    record.set("name", tag.name);
    record.set("slug", tag.slug);
    record.set("color", tag.color);
    record.set("icon", tag.icon);
    record.set("order", tag.order);
    dao.saveRecord(record);
  });

  // 插入初始网站配置
  const configRecord = new Record(siteConfigCollection);
  configRecord.set("site_title", "Pulse Changelog");
  configRecord.set("site_description", "Product update log, recording every step of progress");
  configRecord.set("primary_color", "#3B82F6");
  dao.saveRecord(configRecord);

}, (db) => {
  const dao = new Dao(db);
  // 回滚操作
  try {
    const changelogs = dao.findCollectionByNameOrId("changelogs");
    if (changelogs) dao.deleteCollection(changelogs);
  } catch(e) {}
  try {
    const tags = dao.findCollectionByNameOrId("tags");
    if (tags) dao.deleteCollection(tags);
  } catch(e) {}
  try {
    const siteConfig = dao.findCollectionByNameOrId("site_config");
    if (siteConfig) dao.deleteCollection(siteConfig);
  } catch(e) {}
});

File: migrations/1738223400_initial_schema.go
package migrations

import (
	"github.com/pocketbase/pocketbase/core"
	m "github.com/pocketbase/pocketbase/migrations"
)

func init() {
	m.Register(func(app core.App) error {
		// Create tags collection
		tagsCollection := core.NewBaseCollection("tags")
		tagsCollection.Id = "tags_collection_id"

		tagsCollection.Fields.Add(&core.TextField{
			Name:     "name",
			Required: true,
			Min:      1,
			Max:      50,
		})
		tagsCollection.Fields.Add(&core.TextField{
			Name:     "slug",
			Required: true,
			Min:      1,
			Max:      50,
			Pattern:  "^[a-z0-9]+(?:-[a-z0-9]+)*$",
		})
		tagsCollection.Fields.Add(&core.TextField{
			Name:     "color",
			Required: false,
			Min:      0,
			Max:      7,
		})
		tagsCollection.Fields.Add(&core.TextField{
			Name:     "icon",
			Required: false,
			Min:      0,
			Max:      50,
		})
		tagsCollection.Fields.Add(&core.NumberField{
			Name:     "order",
			Required: false,
			Min:      floatPtr(0),
		})

		tagsCollection.AddIndex("idx_tags_name", true, "name", "")
		tagsCollection.AddIndex("idx_tags_slug", true, "slug", "")

		tagsCollection.ListRule = strPtr("")
		tagsCollection.ViewRule = strPtr("")

		if err := app.Save(tagsCollection); err != nil {
			return err
		}

		// Create changelogs collection
		changelogsCollection := core.NewBaseCollection("changelogs")
		changelogsCollection.Id = "changelogs_collection_id"

		changelogsCollection.Fields.Add(&core.TextField{
			Name:     "title",
			Required: true,
			Min:      1,
			Max:      200,
		})
		changelogsCollection.Fields.Add(&core.EditorField{
			Name:        "description",
			Required:    true,
			ConvertURLs: false,
		})
		changelogsCollection.Fields.Add(&core.TextField{
			Name:     "version",
			Required: true,
			Min:      1,
			Max:      50,
		})
		changelogsCollection.Fields.Add(&core.DateField{
			Name:     "date",
			Required: true,
		})
		changelogsCollection.Fields.Add(&core.RelationField{
			Name:          "tags",
			Required:      false,
			CollectionId:  "tags_collection_id",
			CascadeDelete: false,
		})

		changelogsCollection.AddIndex("idx_changelogs_date", false, "date DESC", "")
		changelogsCollection.AddIndex("idx_changelogs_version", false, "version", "")

		changelogsCollection.ListRule = strPtr("")
		changelogsCollection.ViewRule = strPtr("")

		if err := app.Save(changelogsCollection); err != nil {
			return err
		}

		// Create site_config collection
		siteConfigCollection := core.NewBaseCollection("site_config")
		siteConfigCollection.Id = "site_config_collection_id"

		siteConfigCollection.Fields.Add(&core.TextField{
			Name:     "site_title",
			Required: true,
			Min:      1,
			Max:      100,
		})
		siteConfigCollection.Fields.Add(&core.TextField{
			Name:     "site_description",
			Required: true,
			Min:      1,
			Max:      500,
		})
		siteConfigCollection.Fields.Add(&core.FileField{
			Name:      "logo_url",
			Required:  false,
			MaxSelect: 1,
			MaxSize:   5242880,
			MimeTypes: []string{"image/jpeg", "image/png", "image/svg+xml", "image/gif", "image/webp"},
			Thumbs:    []string{"100x100"},
		})
		siteConfigCollection.Fields.Add(&core.TextField{
			Name:     "primary_color",
			Required: false,
			Min:      0,
			Max:      7,
			Pattern:  "^#[0-9A-Fa-f]{6}$",
		})

		siteConfigCollection.ListRule = strPtr("")
		siteConfigCollection.ViewRule = strPtr("")

		if err := app.Save(siteConfigCollection); err != nil {
			return err
		}

		// Insert initial tags
		initialTags := []struct {
			Name  string
			Slug  string
			Color string
			Icon  string
			Order float64
		}{
			{"New Feature", "new-feature", "#3B82F6", "Sparkles", 1},
			{"Bug Fix", "bug-fix", "#EF4444", "Bug", 2},
			{"Improvement", "improvement", "#10B981", "TrendingUp", 3},
			{"Security Update", "security", "#F59E0B", "Shield", 4},
		}

		for _, tag := range initialTags {
			record := core.NewRecord(tagsCollection)
			record.Set("name", tag.Name)
			record.Set("slug", tag.Slug)
			record.Set("color", tag.Color)
			record.Set("icon", tag.Icon)
			record.Set("order", tag.Order)
			if err := app.Save(record); err != nil {
				return err
			}
		}

		// Insert initial site configuration
		configRecord := core.NewRecord(siteConfigCollection)
		configRecord.Set("site_title", "Changelog")
		configRecord.Set("site_description", "Product update log, recording every step of progress")
		configRecord.Set("primary_color", "#171717")
		if err := app.Save(configRecord); err != nil {
			return err
		}

		return nil
	}, func(app core.App) error {
		// Rollback: delete collections in reverse order
		collections := []string{"changelogs", "tags", "site_config"}
		for _, name := range collections {
			collection, err := app.FindCollectionByNameOrId(name)
			if err == nil && collection != nil {
				if err := app.Delete(collection); err != nil {
					return err
				}
			}
		}
		return nil
	})
}

func strPtr(s string) *string {
	return &s
}

func floatPtr(f float64) *float64 {
	return &f
}
